common/scala/src/main/scala/whisk/core/entity/WhiskPackage.scala

```diff
@@ -140,6 +148,31 @@ object WhiskPackage
     val bindingFieldName = "binding"
     override val collectionName = "packages"
 
+    /**
+     * utility function that traverses all potential binding references and returns the resolved package
+     * @param pkg the package for which the resolution to be applied
+     * @return the same package if there is no binding, or the fully resolved package
+     */
+    def resolveBinding(entityStore: EntityStore, pkg: DocId)(
+        implicit ec: ExecutionContext, transid: TransactionId): Future[WhiskPackage] = {
+        WhiskPackage.get(entityStore, pkg) flatMap {
+            case wp if wp.binding.isEmpty =>
+                Future.successful(wp)
+            case wp => resolveBinding(entityStore, wp.binding.get.docid)
+        }
+    }
+
+    /**
+     * utility function that extracts the package information from a fully qualified name,
+     * which contains a package and returns the corresponding docid
+     * @param fullyQualifiedName the fully qualified name of the entity
+     * @return the docid for the corresponding package
+     */
+    def packageDocId(fullyQualifiedName: String): DocId = {
```

Comment:
leaking the fully qualified names as strings to these entities feels wrong; you don't know if they'll match requirements which means these could fail and throw illegal argument exceptions.

instead the types should be more specific at this point here and resolve them to the narrower types elsewhere; otherwise, handle the non-conformance.